Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,425 @@
+
+section LinearEquiv
+
+variable (e : M ⊗[R] N ≃ₗ[R] R)
+
+open TensorProduct (assoc rid) in
+/-- If M is invertible, `rTensorHom M` admits an inverse. -/
+noncomputable def rTensorInv : (P ⊗[R] M →ₗ[R] Q ⊗[R] M) →ₗ[R] (P →ₗ[R] Q) :=
+  ((assoc R Q M N ≪≫ₗ e.lTensor Q ≪≫ₗ rid R Q).congrRight ≪≫ₗ
+    (assoc R P M N ≪≫ₗ e.lTensor P ≪≫ₗ rid R P).congrLeft _ R) ∘ₗ LinearMap.rTensorHom N
+
+theorem rTensorInv_leftInverse : Function.LeftInverse (rTensorInv P Q e) (.rTensorHom M) :=
+  fun _ ↦ by
+    simp_rw [rTensorInv, LinearEquiv.coe_trans, LinearMap.comp_apply, LinearEquiv.coe_toLinearMap]
+    rw [← LinearEquiv.eq_symm_apply]
+    ext; simp [LinearEquiv.congrLeft, LinearEquiv.congrRight, LinearEquiv.arrowCongrAddEquiv]
+
+theorem rTensorInv_injective : Function.Injective (rTensorInv P Q e) := by
+  simpa [rTensorInv] using (rTensorInv_leftInverse _ _ <| TensorProduct.comm R N M ≪≫ₗ e).injective
+
+/-- If `M` is an invertible `R`-module, `(· ⊗[R] M)` is an auto-equivalence of the category
+of `R`-modules. -/
+@[simps!] noncomputable def rTensorEquiv : (P →ₗ[R] Q) ≃ₗ[R] (P ⊗[R] M →ₗ[R] Q ⊗[R] M) where
+  __ := LinearMap.rTensorHom M
+  invFun := rTensorInv P Q e
+  left_inv := rTensorInv_leftInverse P Q e
+  right_inv _ := rTensorInv_injective P Q e <| by
+    rw [LinearMap.toFun_eq_coe, rTensorInv_leftInverse]
+
+open LinearMap in
+/-- If there is an `R`-isomorphism between `M ⊗[R] N` and `R`, where `M` is not assumed to be
+the dual of `N`, we show the induced map `M → Nᵛ` is in fact an isomorphism, so `M` is
+isomorphic to `Nᵛ` after all. -/
+theorem bijective_curry : Function.Bijective (curry e.toLinearMap) := by
+  have : curry e.toLinearMap = ((TensorProduct.lid R N).congrLeft _ R ≪≫ₗ e.congrRight) ∘ₗ
+      rTensorHom N ∘ₗ (ringLmapEquivSelf R R M).symm.toLinearMap := by
+    rw [← LinearEquiv.toLinearMap_symm_comp_eq]; ext
+    simp [LinearEquiv.congrLeft, LinearEquiv.congrRight, LinearEquiv.arrowCongrAddEquiv]
+  simpa [this] using (rTensorEquiv R M <| TensorProduct.comm R N M ≪≫ₗ e).bijective
+
+include e
+
+protected theorem right : Module.Invertible R N where
+  bijective := by
+    have : contractLeft R N = .symm (.rTensor N (.ofBijective _ <| bijective_curry e)) ≪≫ₗ e := by
+      rw [LinearEquiv.coe_trans, LinearEquiv.eq_comp_toLinearMap_symm]; ext; rfl
+    rw [this]; apply LinearEquiv.bijective
+
+protected theorem left : Module.Invertible R M := .right (TensorProduct.comm R N M ≪≫ₗ e)
+
+instance : Module.Invertible R R := .left (TensorProduct.lid R R)
+
+end LinearEquiv
+
+variable [Module.Invertible R M]
+
+protected theorem congr (e : M ≃ₗ[R] N) : Module.Invertible R N :=
+  .right (e.symm.lTensor _ ≪≫ₗ linearEquiv R M)
+
+variable (R M N)
+
+instance : Module.Invertible R (Dual R M) := .left (linearEquiv R M)
+
+instance [Module.Invertible R N] : Module.Invertible R (M ⊗[R] N) :=
+  .right (M := Dual R M ⊗[R] Dual R N) <| tensorTensorTensorComm .. ≪≫ₗ
+    congr (linearEquiv R M) (linearEquiv R N) ≪≫ₗ TensorProduct.lid R R
```

Comment:
Is it true that if `M` is an invertible `A` module and `N` is an invertible `R` module, then `M ⊗[R] N` is an invertible `A` module?